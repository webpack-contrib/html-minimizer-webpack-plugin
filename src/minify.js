/** @typedef {import("./index.js").MinimizedResult} MinimizedResult */
/** @typedef {import("./index.js").InternalResult} InternalResult */

/**
 * @template T
 * @param {import("./index.js").InternalOptions<T>} options
 * @returns {Promise<InternalResult>}
 */
async function minify(options) {
  const minifyFns = Array.isArray(options.minimizer.implementation)
    ? options.minimizer.implementation
    : [options.minimizer.implementation];

  /** @type {InternalResult} */
  const result = { outputs: [], warnings: [], errors: [] };

  for (let i = 0; i <= minifyFns.length - 1; i++) {
    const minifyFn = minifyFns[i];
    const minifyOptions = Array.isArray(options.minimizer.options)
      ? options.minimizer.options[i]
      : options.minimizer.options;
    const prevResult =
      result.outputs.length > 0
        ? result.outputs[result.outputs.length - 1]
        : { code: options.input };
    const { code } = prevResult;
    /** @type {MinimizedResult} */
    // eslint-disable-next-line no-await-in-loop
    const minifyResult = await minifyFn(
      { [options.name]: code },
      minifyOptions,
    );

    if (
      typeof minifyResult !== "string" &&
      typeof minifyResult.code !== "string"
    ) {
      throw new Error(
        "minimizer function doesn't return the 'code' property or result is not a string value",
      );
    }

    if (typeof minifyResult === "string") {
      result.outputs.push({ code: minifyResult });
    } else {
      if (minifyResult.errors) {
        result.errors = result.errors.concat(minifyResult.errors);
      }

      if (minifyResult.warnings) {
        result.warnings = result.warnings.concat(minifyResult.warnings);
      }

      result.outputs.push({ code: minifyResult.code });
    }
  }

  result.outputs = [result.outputs[result.outputs.length - 1]];

  return result;
}

/**
 * @template T
 * @param {import("./index.js").InternalOptions<T>} options
 * @returns {Promise<InternalResult>}
 */
async function transform(options) {
  // 'use strict' => this === undefined (Clean Scope)
  // Safer for possible security issues, albeit not critical at all here
  // eslint-disable-next-line no-new-func, no-param-reassign
  const evaluatedOptions = new Function(
    "exports",
    "require",
    "module",
    "__filename",
    "__dirname",
    `'use strict'\nreturn ${options}`,
  )(exports, require, module, __filename, __dirname);

  return minify(evaluatedOptions);
}

module.exports = { minify, transform };
