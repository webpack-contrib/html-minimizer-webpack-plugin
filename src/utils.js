/* istanbul ignore next */
async function htmlMinifierTerser(data, minimizerOptions) {
  // eslint-disable-next-line global-require,import/no-extraneous-dependencies
  const htmlMinifier = require("html-minifier-terser");

  const defaultMinimizerOptions = {
    caseSensitive: true,
    // `collapseBooleanAttributes` is not always safe, since this can break CSS attribute selectors and not safe for XHTML
    collapseWhitespace: true,
    conservativeCollapse: true,
    keepClosingSlash: true,
    // We need ability to use cssnano, or setup own function without extra dependencies
    minifyCSS: true,
    minifyJS: true,
    // `minifyURLs` is unsafe, because we can't guarantee what the base URL is
    // `removeAttributeQuotes` is not safe in some rare cases, also HTML spec recommends against doing this
    removeComments: true,
    // `removeEmptyAttributes` is not safe, can affect certain style or script behavior, look at https://github.com/webpack-contrib/html-loader/issues/323
    // `removeRedundantAttributes` is not safe, can affect certain style or script behavior, look at https://github.com/webpack-contrib/html-loader/issues/323
    removeScriptTypeAttributes: true,
    removeStyleLinkTypeAttributes: true,
    // `useShortDoctype` is not safe for XHTML
  };

  const [[, input]] = Object.entries(data);
  const result = await htmlMinifier.minify(input, {
    ...defaultMinimizerOptions,
    ...minimizerOptions,
  });

  return { code: result };
}

// eslint-disable-next-line import/prefer-default-export
export { htmlMinifierTerser };
