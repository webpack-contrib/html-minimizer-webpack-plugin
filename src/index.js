import os from "os";

import { validate } from "schema-utils";
import serialize from "serialize-javascript";
import { Worker } from "jest-worker";

import schema from "./options.json";

import { htmlMinifierTerser, throttleAll } from "./utils";
import { minify as minifyFn } from "./minify";

class HtmlMinimizerPlugin {
  constructor(options = {}) {
    validate(schema, options, {
      name: "Html Minimizer Plugin",
      baseDataPath: "options",
    });

    const {
      minify = htmlMinifierTerser,
      minimizerOptions,
      test = /\.html(\?.*)?$/i,
      parallel = true,
      include,
      exclude,
    } = options;

    this.options = {
      test,
      parallel,
      include,
      exclude,
      minify,
      minimizerOptions,
    };
  }

  static buildWarning(warning, file) {
    const builtWarning = new Error(
      warning.message ? warning.message : warning.toString()
    );

    builtWarning.name = "Warning";
    builtWarning.hideStack = true;
    builtWarning.file = file;

    return builtWarning;
  }

  static buildError(error, file) {
    let builtError;

    if (typeof error === "string") {
      // @ts-ignore
      builtError = new Error(`${file} from Html Minimizer plugin\n${error}`);
      builtError.file = file;

      return builtError;
    }

    if (error.stack) {
      // @ts-ignore
      builtError = new Error(
        `${file} from Html Minimizer plugin\n${
          typeof error.message !== "undefined" ? error.message : ""
        }\n${error.stack}`
      );
      builtError.file = file;

      return builtError;
    }

    builtError = new Error(
      `${file} from Html Minimizer plugin\n${error.message}`
    );
    builtError.file = file;

    return builtError;
  }

  static getAvailableNumberOfCores(parallel) {
    // In some cases cpus() returns undefined
    // https://github.com/nodejs/node/issues/19022
    const cpus = os.cpus() || { length: 1 };

    return parallel === true
      ? cpus.length - 1
      : Math.min(Number(parallel) || 0, cpus.length - 1);
  }

  async optimize(compiler, compilation, assets, optimizeOptions) {
    const cache = compilation.getCache("HtmlMinimizerWebpackPlugin");
    let numberOfAssets = 0;
    const assetsForMinify = await Promise.all(
      Object.keys(assets)
        .filter((name) => {
          const { info } = compilation.getAsset(name);

          // Skip double minimize assets from child compilation
          if (info.minimized) {
            return false;
          }

          if (
            !compiler.webpack.ModuleFilenameHelpers.matchObject.bind(
              // eslint-disable-next-line no-undefined
              undefined,
              this.options
            )(name)
          ) {
            return false;
          }

          return true;
        })
        .map(async (name) => {
          const { info, source } = compilation.getAsset(name);

          const eTag = cache.getLazyHashedEtag(source);
          const cacheItem = cache.getItemCache(name, eTag);
          const output = await cacheItem.getPromise();

          if (!output) {
            numberOfAssets += 1;
          }

          return { name, info, inputSource: source, output, cacheItem };
        })
    );

    if (assetsForMinify.length === 0) {
      return;
    }

    let getWorker;
    let initializedWorker;
    let numberOfWorkers;

    if (optimizeOptions.availableNumberOfCores > 0) {
      // Do not create unnecessary workers when the number of files is less than the available cores, it saves memory
      numberOfWorkers = Math.min(
        numberOfAssets,
        optimizeOptions.availableNumberOfCores
      );
      // eslint-disable-next-line consistent-return
      getWorker = () => {
        if (initializedWorker) {
          return initializedWorker;
        }

        initializedWorker = new Worker(require.resolve("./minify"), {
          numWorkers: numberOfWorkers,
          enableWorkerThreads: true,
        });

        // https://github.com/facebook/jest/issues/8872#issuecomment-524822081
        const workerStdout = initializedWorker.getStdout();

        if (workerStdout) {
          workerStdout.on("data", (chunk) => process.stdout.write(chunk));
        }

        const workerStderr = initializedWorker.getStderr();

        if (workerStderr) {
          workerStderr.on("data", (chunk) => process.stderr.write(chunk));
        }

        return initializedWorker;
      };
    }

    const { RawSource } = compiler.webpack.sources;
    const scheduledTasks = [];

    for (const asset of assetsForMinify) {
      scheduledTasks.push(async () => {
        const { name, inputSource, cacheItem } = asset;
        let { output } = asset;
        let input;

        const sourceFromInputSource = inputSource.source();

        if (!output) {
          input = sourceFromInputSource;

          if (Buffer.isBuffer(input)) {
            input = input.toString();
          }

          const options = {
            name,
            input,
            minimizerOptions: this.options.minimizerOptions,
            minify: this.options.minify,
          };

          try {
            output = await (getWorker
              ? getWorker().transform(serialize(options))
              : minifyFn(options));
          } catch (error) {
            compilation.errors.push(
              HtmlMinimizerPlugin.buildError(error, name)
            );

            return;
          }

          output.source = new RawSource(output.code);

          await cacheItem.storePromise({
            source: output.source,
            errors: output.errors,
            warnings: output.warnings,
          });
        }

        const newInfo = { minimized: true };

        if (output.warnings && output.warnings.length > 0) {
          for (const warning of output.warnings) {
            compilation.warnings.push(
              HtmlMinimizerPlugin.buildWarning(warning, name)
            );
          }
        }

        if (output.errors && output.errors.length > 0) {
          for (const error of output.errors) {
            compilation.errors.push(
              HtmlMinimizerPlugin.buildError(error, name)
            );
          }
        }

        compilation.updateAsset(name, output.source, newInfo);
      });
    }

    const limit =
      getWorker && numberOfAssets > 0
        ? /** @type {number} */ (numberOfWorkers)
        : scheduledTasks.length;

    await throttleAll(limit, scheduledTasks);

    if (initializedWorker) {
      await initializedWorker.end();
    }
  }

  apply(compiler) {
    const pluginName = this.constructor.name;
    const availableNumberOfCores =
      HtmlMinimizerPlugin.getAvailableNumberOfCores(this.options.parallel);

    compiler.hooks.compilation.tap(pluginName, (compilation) => {
      compilation.hooks.processAssets.tapPromise(
        {
          name: pluginName,
          stage:
            compiler.webpack.Compilation.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
          additionalAssets: true,
        },
        (assets) =>
          this.optimize(compiler, compilation, assets, {
            availableNumberOfCores,
          })
      );

      compilation.hooks.statsPrinter.tap(pluginName, (stats) => {
        stats.hooks.print
          .for("asset.info.minimized")
          .tap(
            "html-minimizer-webpack-plugin",
            (minimized, { green, formatFlag }) =>
              // eslint-disable-next-line no-undefined
              minimized ? green(formatFlag("minimized")) : undefined
          );
      });
    });
  }
}

HtmlMinimizerPlugin.htmlMinifierTerser = htmlMinifierTerser;

export default HtmlMinimizerPlugin;
