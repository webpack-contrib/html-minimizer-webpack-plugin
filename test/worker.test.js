import serialize from "serialize-javascript";

import { transform } from "../src/minify";
import HtmlMinimizerPlugin from "../src";

import { normalizeErrors } from "./helpers";

describe("worker", () => {
  it("should minify html", async () => {
    const options = {
      name: "entry.html",
      input: '<!-- Comment --><p title="blah" id="moo">     foo     </p>',
      minimizer: {
        implementation: HtmlMinimizerPlugin.htmlMinifierTerser,
        options: {
          removeComments: false,
        },
      },
    };
    const { code } = await transform(serialize(options));

    expect(code).toMatchSnapshot("html");
  });

  it("should minify html #2", async () => {
    const options = {
      name: "entry.html",
      input: '<!-- Comment --><p title="blah" id="moo">     foo     </p>',
      minimizer: [
        {
          implementation: HtmlMinimizerPlugin.htmlMinifierTerser,
          options: {
            removeComments: false,
          },
        },
      ],
    };
    const { code } = await transform(serialize(options));

    expect(code).toMatchSnapshot("html");
  });

  it("should minify html #3", async () => {
    const options = {
      name: "entry.html",
      input: '<!-- Comment --><p title="blah" id="moo">     foo     </p>',
      minimizer: {
        implementation: () => {
          return {
            html: '<!-- From minify function --><p class="atata">from-minify-function</p>',
          };
        },
        options: { removeComments: false },
      },
    };
    const { code } = await transform(serialize(options));

    expect(code).toMatchSnapshot("html");
  });

  it("should emit error", async () => {
    const options = {
      name: "entry.html",
      input: false,
      minimizer: {
        implementation: HtmlMinimizerPlugin.htmlMinifierTerser,
      },
    };

    try {
      await transform(serialize(options));
    } catch (error) {
      const normalizeError = { ...error };

      normalizeError.message = [error.message.split("\n")];

      expect(normalizeErrors(normalizeError.message)).toMatchSnapshot("error");
    }
  });
});
