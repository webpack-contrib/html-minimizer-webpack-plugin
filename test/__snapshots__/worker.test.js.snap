// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`worker should emit error: error 1`] = `
Array [
  "str.replace is not a function",
]
`;

exports[`worker should minify html #2: html 1`] = `"<!-- Comment --><p title=\\"blah\\" id=\\"moo\\"> foo </p>"`;

exports[`worker should minify html #3: html 1`] = `
Object {
  "html": "<!-- From minify function --><p class=\\"atata\\">from-minify-function</p>",
}
`;

exports[`worker should minify html: html 1`] = `"<!-- Comment --><p title=\\"blah\\" id=\\"moo\\"> foo </p>"`;
